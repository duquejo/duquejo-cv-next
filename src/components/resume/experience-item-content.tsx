import { cn } from '@/lib';
import { Button } from '@/components/ui/button';
import { ExperienceType } from '@/interfaces';
import Link from 'next/link';
import Image from 'next/image';

interface Props extends Partial<ExperienceType> {
  contentTitle?: string;
  button?: string;
}

export const ExperienceItemContent = ({
  project = 'Image',
  resume,
  button,
  screenshot,
  url,
  contentTitle,
}: Props) => {
  return (
    <div className="flex flex-col lg:flex-row gap-x-4">
      {resume && (
        <div
          className={cn(
            'flex flex-col order-2 lg:order-1 justify-between items-baseline',
            screenshot && 'basis-8/12',
          )}
        >
          <div className="flex-auto">
            {contentTitle && <span className="font-semibold text-base">{contentTitle}</span>}
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-2 text-xs text-justify leading-normal">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-xs leading-normal">{resume}</p>
            )}
          </div>
          {url && button && (
            <Button
              size="sm"
              className="mt-5 lg:flex-shrink lg:w-auto w-full"
              variant="secondary"
              asChild
            >
              <Link href={url} target="_blank" rel="noopener noreferrer nofollow">
                {button}
              </Link>
            </Button>
          )}
        </div>
      )}
      {screenshot && (
        <picture className="order-1 lg:order-2 flex flex-grow flex-col items-center px-0 mb-5 pt-0 lg:px-4 lg:pt-8 lg:mb-0 lg:max-h-60 lg:max-w-xs overflow-hidden">
          <Image
            src={screenshot}
            width="300"
            height="150"
            loading="lazy"
            alt={project}
            className="rounded object-cover shadow-lg w-auto"
          />
        </picture>
      )}
    </div>
  );
};
