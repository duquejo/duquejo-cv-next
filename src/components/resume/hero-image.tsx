import Image from 'next/image';
import { cn } from '@/lib';
import { Circle, Gamepad, LoaderCircle, Music, Send } from 'lucide-react';
import avatarPicture from '../../../public/static/img/avatar_410.png';

interface Props {
  className?: string;
}

export const HeroImage = ({ className }: Props) => {
  return (
    <div data-testid="hero-image" className={cn('overflow-hidden relative', className)}>
      <Music
        className="absolute top-0 text-primary/50 animate-bounce-subtle repeat-infinite z-10"
        strokeWidth="3"
        size={54}
      />
      <Gamepad
        className="absolute bottom-0 left-7 text-primary/50 animate-wiggle z-10"
        strokeWidth="3"
        size={60}
      />
      <LoaderCircle
        className="absolute right-10 text-primary animate-spin duration-slow repeat-infinite z-10"
        strokeWidth="5"
        size={44}
      />
      <Send
        className="absolute right-8 bottom-5 text-primary animate-bounce-subtle repeat-infinite z-10"
        strokeWidth="3"
        size={50}
      />

      {/*Right*/}
      <Circle
        className="absolute right-36 top-10 text-primary/50 animate-pulse duration-slow"
        strokeWidth="5"
        size={40}
      />
      {/*Left*/}
      <Circle
        className="absolute left-20 bottom-24 text-primary/50 animate-pulse duration-slow delay-150"
        strokeWidth="5"
        size={20}
      />
      <picture className="relative">
        <Image
          className="rounded-full backdrop-blur-sm bg-primary/10 shadow-inner"
          alt="JosÃ© Miguel Duque"
          height={410}
          width={410}
          priority={true}
          quality={100}
          src={avatarPicture}
        />
      </picture>
    </div>
  );
};
