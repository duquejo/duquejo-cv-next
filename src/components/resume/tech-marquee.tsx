import { Marquee } from '@/components/ui/marquee';
import { STACK_DATA as stack } from '@/lib';
import { Badge } from '@/components/ui/badge';
import Image from 'next/image';

export const TechMarquee = () => {
  return (
    <Marquee className="[--duration:30s]">
      {stack.map((s) => (
        <Badge
          key={s.title}
          variant="secondary"
          className="flex gap-x-2 px-1.5 py-1 lg:p-1.5 rounded-xl"
        >
          <Image src={s.icon} alt={s.title} width={20} height={20} className="max-w-none" />
          {s.title}
        </Badge>
      ))}
    </Marquee>
  );
};
