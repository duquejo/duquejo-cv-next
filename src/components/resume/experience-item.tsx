import { cn } from '@/lib';
import type { ExperienceType } from '@/interfaces';

import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import { ExperienceItemContent } from '@/components/resume/experience-item-content';

interface Props extends ExperienceType {
  className?: string;
  experienceItemTitle?: string;
  experienceItemButtonLabel?: string;
  experienceFooterText?: string;
}

export const ExperienceItem = ({
  className,
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
  screenshot,
  experienceItemTitle,
  experienceItemButtonLabel,
  experienceFooterText,
  url,
}: Props) => {
  return (
    <li
      className={cn(
        'first-of-type:pt-0 pt-5 relative border-s border-gray-200 list-none',
        className,
      )}
    >
      <span
        data-testid="detail"
        className="absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-primary"
      />
      <span className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-testid="time"
          className={cn(
            'lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-primary' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          {end_date && (
            <time className="block lg:before:content-['-'] lg:before:mx-1">{end_date}</time>
          )}
        </span>
        {enterprise && (
          <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
        )}
      </span>
      <Card className="ms-4 bg-sidebar/30 border-dashed">
        <CardHeader>
          <CardTitle>
            <h2>{project}</h2>
          </CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        <CardContent className="py-0">
          <ExperienceItemContent
            contentTitle={experienceItemTitle}
            button={experienceItemButtonLabel}
            project={project}
            resume={resume}
            screenshot={screenshot}
            url={url}
          />
        </CardContent>
        {additional_info && (
          <CardFooter className="flex flex-col items-start gap-y-2 pt-5">
            <span className="font-semibold text-base">{experienceFooterText}</span>
            <div className="flex lg:gap-1 gap-2 flex-wrap lg:justify-normal justify-around">
              {additional_info.map((info) => (
                <Badge variant="secondary" className="cursor-pointer" key={info}>
                  {info}
                </Badge>
              ))}
            </div>
          </CardFooter>
        )}
      </Card>
    </li>
  );
};
