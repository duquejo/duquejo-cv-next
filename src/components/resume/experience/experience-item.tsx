import { cn } from '@/lib/utils';
import type { ExperienceType } from '@/interfaces/experience.interface';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import { ExperienceItemContent } from '@/components/resume/experience/experience-item-content';

export const ExperienceItem = ({
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
  screenshot,
  url,
}: ExperienceType) => {
  return (
    <ol className="first-of-type:pt-0 pt-5 relative border-s border-gray-200">
      <div
        data-testid="detail"
        className={cn(
          'absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400',
          isRecent ? 'bg-primary' : 'bg-secondary',
        )}
      />

      {/*Aligned content*/}
      <div className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-testid="time"
          className={cn(
            'lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-primary' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          {end_date && (
            <time className="block lg:before:content-['-'] lg:before:mx-1">{end_date}</time>
          )}
        </span>
        {enterprise && (
          <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
        )}
      </div>
      <Card className="ms-4">
        <CardHeader>
          <CardTitle>{project}</CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        <CardContent className="py-0">
          <ExperienceItemContent
            project={project}
            resume={resume}
            screenshot={screenshot}
            url={url}
          />
        </CardContent>
        {additional_info && (
          <CardFooter className="flex flex-col items-start gap-y-2 pt-5">
            <div className="font-semibold text-base">Skills</div>
            <div className="flex lg:gap-1 gap-2 flex-wrap lg:justify-normal justify-around">
              {additional_info.map((info) => (
                <Badge variant="secondary" key={info}>
                  {info}
                </Badge>
              ))}
            </div>
          </CardFooter>
        )}
      </Card>
    </ol>
  );
};
