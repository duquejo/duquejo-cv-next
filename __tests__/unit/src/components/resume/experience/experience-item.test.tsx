import type { ExperienceType } from '@/interfaces';
import { render, screen } from '@testing-library/react';
import { ExperienceItem } from '@/components/resume/experience-item';

vi.mock('next-intl', () => ({
  useTranslations: () => (key: string) =>
    ({
      subtitle: 'subtitle',
      button: 'button',
      footer: 'footer',
    })[key],
}));

describe('<ExperienceItemContent /> tests', () => {
  const data: Required<ExperienceType> = {
    resume: ['foo', 'bar', 'baz'],
    screenshot: '/foo/bar.webp',
    url: 'https://foo.bar',
    enterprise: 'John Doe',
    additional_info: ['Next JS', 'React', 'Vitest'],
    isRecent: false,
    start_date: '2025-01-01',
    end_date: '2025-01-02',
    role: 'Software analyst',
    project: 'Foo-Bar Baz Inc.',
  };

  it('Should match the snapshot with the default args', () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { resume, additional_info, screenshot, url, end_date, ...mandatoryFields } = data;

    const { container } = render(<ExperienceItem {...mandatoryFields} />);

    expect(container).toMatchSnapshot();

    expect(screen.getByTestId('time')).not.toHaveTextContent(end_date);
    expect(screen.getByTestId('time')).not.toHaveClass('border-yellow-400');

    expect(screen.getByTestId('detail')).not.toHaveClass('bg-primary');
  });

  it('Should match the snapshot with the available args', () => {
    const { container } = render(<ExperienceItem {...data} />);

    expect(container).toMatchSnapshot();

    expect(screen.getByText('subtitle')).toBeInTheDocument();
    data.additional_info.forEach((item: string) =>
      expect(screen.getByText(item)).toBeInTheDocument(),
    );

    expect(screen.getByTestId('time')).toHaveTextContent(data.start_date);
    expect(screen.getByTestId('time')).toHaveTextContent(data.end_date);
  });

  it('Should stylize the time tags if the parameter "isRecent" is truthy', () => {
    const isRecentWithData = {
      ...data,
      isRecent: true,
    };

    render(<ExperienceItem {...isRecentWithData} />);

    expect(screen.getByTestId('time')).toHaveClass('border-primary');
    expect(screen.getByTestId('detail')).toHaveClass('bg-primary');
  });
});
