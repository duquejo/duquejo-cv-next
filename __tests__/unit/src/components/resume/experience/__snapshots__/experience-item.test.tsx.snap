// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ExperienceItemContent /> tests > Should match the snapshot with the available args 1`] = `
<div>
  <li
    class="first-of-type:pt-0 pt-5 relative border-s border-gray-200 list-none"
  >
    <span
      class="absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400, bg-primary"
      data-testid="detail"
    />
    <span
      class="pb-2 ms-4 flex items-center justify-between lg:justify-normal"
    >
      <span
        class="lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center border-secondary"
        data-testid="time"
      >
        <time
          class="block"
        >
          2025-01-01
        </time>
        <time
          class="block lg:before:content-['-'] lg:before:mx-1"
        >
          2025-01-02
        </time>
      </span>
      <span
        class="text-xs font-normal ml-2 text-right lg:text-left"
      >
        John Doe
      </span>
    </span>
    <div
      class="rounded-xl border bg-card text-card-foreground shadow ms-4"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <div
          class="font-semibold leading-none tracking-tight"
        >
          <h2>
            Foo-Bar Baz Inc.
          </h2>
        </div>
        <div
          class="text-sm text-muted-foreground"
        >
          Software analyst
        </div>
      </div>
      <div
        class="p-6 py-0"
      >
        <div
          class="flex flex-col lg:flex-row gap-x-4"
        >
          <div
            class="flex flex-col order-2 lg:order-1 justify-between items-baseline basis-8/12"
          >
            <div
              class="flex-auto"
            >
              <span
                class="font-semibold text-base"
              >
                experience item title
              </span>
              <ol
                class="list-disc ml-4 text-justify marker:text-secondary"
              >
                <li
                  class="my-2 text-xs text-justify leading-normal"
                >
                  foo
                </li>
                <li
                  class="my-2 text-xs text-justify leading-normal"
                >
                  bar
                </li>
                <li
                  class="my-2 text-xs text-justify leading-normal"
                >
                  baz
                </li>
              </ol>
            </div>
            <a
              class="inline-flex items-center justify-center gap-2 whitespace-nowrap font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-8 rounded-md px-3 text-xs mt-5 lg:flex-shrink lg:w-auto w-full"
              href="https://foo.bar"
              rel="noopener noreferrer nofollow"
              target="_blank"
            >
              experience button label
            </a>
          </div>
          <picture
            class="order-1 lg:order-2 flex flex-grow flex-col items-center px-0 mb-5 pt-0 lg:px-4 lg:pt-8 lg:mb-0 lg:max-h-60 lg:max-w-xs overflow-hidden"
          >
            <img
              alt="Foo-Bar Baz Inc."
              class="rounded object-cover shadow-lg w-auto"
              data-nimg="1"
              decoding="async"
              height="150"
              loading="lazy"
              src="/_next/image?url=%2Ffoo%2Fbar.webp&w=640&q=75"
              srcset="/_next/image?url=%2Ffoo%2Fbar.webp&w=384&q=75 1x, /_next/image?url=%2Ffoo%2Fbar.webp&w=640&q=75 2x"
              style="color: transparent;"
              width="300"
            />
          </picture>
        </div>
      </div>
      <div
        class="p-6 flex flex-col items-start gap-y-2 pt-5"
      >
        <span
          class="font-semibold text-base"
        >
          experience footer text
        </span>
        <div
          class="flex lg:gap-1 gap-2 flex-wrap lg:justify-normal justify-around"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 cursor-pointer"
          >
            Next JS
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 cursor-pointer"
          >
            React
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 cursor-pointer"
          >
            Vitest
          </div>
        </div>
      </div>
    </div>
  </li>
</div>
`;

exports[`<ExperienceItemContent /> tests > Should match the snapshot with the default args 1`] = `
<div>
  <li
    class="first-of-type:pt-0 pt-5 relative border-s border-gray-200 list-none"
  >
    <span
      class="absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400, bg-primary"
      data-testid="detail"
    />
    <span
      class="pb-2 ms-4 flex items-center justify-between lg:justify-normal"
    >
      <span
        class="lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center border-secondary"
        data-testid="time"
      >
        <time
          class="block"
        >
          2025-01-01
        </time>
      </span>
      <span
        class="text-xs font-normal ml-2 text-right lg:text-left"
      >
        John Doe
      </span>
    </span>
    <div
      class="rounded-xl border bg-card text-card-foreground shadow ms-4"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <div
          class="font-semibold leading-none tracking-tight"
        >
          <h2>
            Foo-Bar Baz Inc.
          </h2>
        </div>
        <div
          class="text-sm text-muted-foreground"
        >
          Software analyst
        </div>
      </div>
      <div
        class="p-6 py-0"
      >
        <div
          class="flex flex-col lg:flex-row gap-x-4"
        />
      </div>
      <div
        class="p-6 flex flex-col items-start gap-y-2 pt-5"
      >
        <span
          class="font-semibold text-base"
        >
          experience footer text
        </span>
        <div
          class="flex lg:gap-1 gap-2 flex-wrap lg:justify-normal justify-around"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 cursor-pointer"
          >
            Next JS
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 cursor-pointer"
          >
            React
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80 cursor-pointer"
          >
            Vitest
          </div>
        </div>
      </div>
    </div>
  </li>
</div>
`;
